docs/adr/0011-brighter-support-for-bulk-messaging-operations.md

```diff
@@ -0,0 +1,86 @@
+# 11. Support for Bulk Messaging Operations 
+
+Date: 2019-08-01
+
+## Status
+
+Proposed
+
+## Context
+   
+To support Transactional Messaging the Command Processor uses an Outbox when producing a message via an external bus. This means that the CommandProcessor has two associated calls:
+
+* A `DepositPost` call to translate the `IRequest` into a `Message` and store it in an `Outbox`
+* A `ClearOutbox` call to retrieve the `Message` from the `Outbox` and dispatch it via a `Producer`
+
+Within these two steps we peform a number of activities.
+
+For `DepositPost` we:
+
+* Lookup the `IAmAProducer` associated with a `Request
+* Translate an `IRequest` into the `Message`, excuting the `IAmAMessageTransform` and `IAmAMessagemapper` pipeline
+* Store the resulting `Message` in an `Outbox`
+
+For `ClearOutbox` we:
+
+* Retrieve the `Message` from the `Outbox`
+* We then produce the `Message` via the `Producer`
+
+For the sake of efficiency we chose to support passing an array of `IRequest` to `DepositPost`. We want to bulk write this set of messages to the Outbox, over writing each one individually. For this reason we would like to be able to batch up our writes to the `Outbox` in `DepositPost`
+
+We always pass an array of `Message` identifiers to `ClearOutbox`. For efficiency we obtain the set of matching ids from the `Outbox`
+
+A question arises from the relationship between the single `DepositPost` and the bulk `DepositPost`.
+
+The bulk version of `DepositPost` requires two modfications to the deposit behavior:
+
+* As it is an `IEnumerable<IRequest>` the collection passed to the bulk `DepositPost` does not bind a generic parameter of the derived type of `IRequest` that we can use to lookup the associated mapper/transform pipeline and producer.
+* As it is a batch, we want to accumulate all the writes to the `Outbox` over making each one individually.
+
+In addition though we would like to re-use the `Deposit Post` functionality as far as possible to avoid divergent code paths reducing modifiability such as when adding OTel or fixing bugs. 
+
+In V9 our solution was as follows:
+
+* Bucket any batch of `IRequest` by their type
+* By batch late bind a `BulkMessageMap` and then run it through the collection
+* Feed the resulting set of messages to a bulk version of the `Outbox` via `AddToOutbox`
+
+The main issue with this approach is that the flow for `DepositPost` differs between bulk and single versions, which makes it more complex to modify. 
+
+## Decision
+
+It is worth noting the following:
+
+* The `BulkMessageMap` is a bulk, iterative operation; it's main purpose is late-binding of the type from an `IEnumerable<IRequest>`
+* Producing a `Message` via a `Producer` is not a bulk operation in Brighter
```

Comment:
Brighter v9 does support Bulk clearing via the `IAmABulkMessageProducerAsync` interface

Reply:
I'll take a look, although we have not made changes to clear along these lines yet